pkg/containerdutil/content.go

```diff
@@ -0,0 +1,75 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+// Package containerdutil provides "caching" versions of containerd native snapshotter and content store.
+// NOTE: caching should only be used for single, atomic operations, like `nerdctl images`, and NOT kept
+// across successive, distincts operations. As such, caching is not persistent across invocations of nerdctl,
+// and only lasts as long as the lifetime of the Snapshotter or ContentStore.
+package containerdutil
+
+import (
+	"context"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/content"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type Provider = content.Provider
+type ReaderAt = content.ReaderAt
+
+// ContentStore should be called to get a Provider with caching
+func ContentStore(client *containerd.Client) Provider {
+	return &providerWithCache{
+		client.ContentStore(),
+		make(map[string]content.ReaderAt),
+	}
+}
+
+type providerWithCache struct {
+	native content.Provider
+	cache  map[string]content.ReaderAt
```

Comment:
Closed `RedaerAt` shouldn't be reused. I don't think it's reusable.

Reply:
In that case, given the code paths we are hitting here, I believe this is fine.
However, agreed it is risky, and you are right.
Next push will change that so we get a new ReaderAt every time - while still caching size.